common/txmgr/types/tx.go

```diff
@@ -0,0 +1,21 @@
+package types
+
+import (
+	uuid "github.com/satori/go.uuid"
+	"github.com/smartcontractkit/chainlink/core/services/pg"
+)
+
+// TxStrategy controls how txes are queued and sent
+//
+//go:generate mockery --quiet --name TxStrategy --output ./mocks/ --case=underscore --structname TxStrategy --filename tx_strategy.go
+type TxStrategy interface {
+	// Subject will be saved txes.subject if not null
+	Subject() uuid.NullUUID
+	// PruneQueue is called after tx insertion
+	// It accepts the service responsible for deleting
+	// unstarted txs and deletion options
+	PruneQueue(pruneService UnstartedTxQueuePruner, qopt pg.QOpt) (n int64, err error)
+}
+
+// No constants are defined because states may vary based on chain
+type TxAttemptState = string
```

Comment:
I also made `type EthTxAttemptState = string` for this to work. This was the easiest way to introduce TxAttemptState basically by saying that TxAttemptState is just a string.

Other alternative is to enforce a strict definition and basically replace  `type EthTxAttemptState = string` with `type TxAttemptState string`. In this approach it is not clear how to extend chain states if necessary so I'm not sure which one is better.

Reply:
What is the advantage of defining types if they are only going to be pure strings? Is that better than just using strings?

I would think that for this to be extensible for each chain family, we would need generics, in which case we could use `~string` type constraints to restrict the underlying type while still allowing each chain to define their own custom type with constant vals to use as an 'enum'.

Reply:
I'd rather see us keep the custom type. We are losing type-safety this way.

Reply:
All of these different options would preserve type safety and the ability to do direct conversions `string(var)`:
```go
type CustomString string
type AliasString = string
type Constraint ~string
type Constraint interface {
	~string
	CustomMethod() []byte
}
```
Using generics with constraints makes it extensible with standard go idioms. Using a single custom type or alias does not.

Reply:
> I'd rather see us keep the custom type. We are losing type-safety this way.

After thinking about it, if we keep one custom type (ex: `type TxAttemptState string`), when we add states that one chain has that another doesn't, we are also losing type-safety because you could use states that are invalid for one chain.

I think I like Jordan's approach: each chain can define their own custom types so you don't lose type safety. For states that are equivalent and have shared logic in the generalized TXM you can convert to string safely as well.

Reply:
TxAttempt states are core part of Txmgr, and common for all chains.
For some chains, state transitions may make certain states invalid, that should be ok.
Also, I don't see any reason why we should re-imagine this TxAttemptState in this PR.

Why can't we keep it untouched for now? And assume all chains will use the same common states?

If we realize in future that chains may need custom state, we will have to think broader than just this type. With custom states, we will also need new chain-specific modules which perform those state transitions. Thus I feel we should just leave this untouched now.

Reply:
If we don't intend to use custom states, then I agree it is premature to worry about making this generic/extensible. However, having a bunch of extra states in a universal bucket that are ignored by some chains is equivalent so I don't follow the rationale - it sounds like we already have custom states effectively built-in to the model, why not try to capture that pattern in a way that will scale?

Reply:
Ok, seems like everyone is good with idea to keep custom types as is for now. We can revisit at a later date when we come across chain-specific tx attempt states.